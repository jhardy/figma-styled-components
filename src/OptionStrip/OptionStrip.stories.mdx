import { OptionStrip } from './OptionStrip'
import { Story, Meta } from '@storybook/addon-docs/blocks'

<Meta title='MDX|OptionStrip' component={OptionStrip} />

# OptionStrip

<Story name='OptionStrip'>
  <div className='component-box'>
    <OptionStrip
      defaultValue={'2'}
      options={[
        { value: '1', label: 'x' },
        { value: '2', label: 'y' },
        { value: '3', label: 'z' }
      ]}
    />
  </div>
</Story>

<div className="props-table">

## Props

| Prop Name    | Type                                                                  | Default | Required | Description                              |
| :----------- | :-------------------------------------------------------------------- | :-----: | :------: | :--------------------------------------- |
| options      | <code>Array<{value: string, label: JSX.Element &#124; string}></code> |    -    |   yes    | Array of options for the strip           |
| defaultValue | String                                                                |    -    |    -     | Default Value                            |
| value        | String                                                                |    -    |    -     | Value stored in state of component       |
| onChange     | function                                                              |    -    |    -     | Callback for when an options has changed |

</div>
