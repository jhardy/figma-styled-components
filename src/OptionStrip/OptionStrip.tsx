import * as React from 'react'
import styled from 'styled-components'
import { Text } from '../Text'

/* TODO Type label better to accept, strings, numbers, and SVG */
export interface StripOption {
  value: string
  label: JSX.Element | string
}


// TODO figure out why onchange even fires twice here
export interface OptionStripProps {
  options: StripOption[]
  defaultValue?: string
  value?: string
  onChange?: (event: React.ChangeEvent<HTMLInputElement>) => void
}

const OptionStripContainer = styled.div`
  box-sizing: border-box;
  border-radius: 2px;
  display: inline-flex;
  height: 24px;
  justify-content: center;

  &:hover {
    box-shadow: inset 0 0 0 1px rgba(0, 0, 0, 0.1);
  }
`

const OptionStripOption = styled.div<{ selected?: boolean }>`
  align-items: center;
  background: ${(props) => (props.selected ? 'rgba(0,0,0,0.1)' : 'transparent')};
  box-sizing: border-box;
  border-radius: 2px;
  display: inline-flex;
  flex-grow: 1;
  height: 24px;
  justify-content: center;
  min-width: 24px;
  padding: 5px;
  position: relative;

  ${Text} {
    line-height: 0;
  }

  input {
    position: absolute;
    width: 100%;
    height: 100%;
    opacity: 0;
  }
`

class OptionStripFactory extends React.Component<
  OptionStripProps,
  { options: StripOption[]; defaultSelected: string; value: string }
> {
  constructor (props: OptionStripProps) {
    super(props)
    this.state = {
      defaultSelected: this.props.defaultValue || this.props.options[0].value,
      options: this.props.options,
      value: this.props.value ? this.props.value : this.props.defaultValue ? this.props.defaultValue : this.props.options[0].value
    }
    this.handleChange = this.handleChange.bind(this)
  }

  public render () {
    return (
      <OptionStripContainer {...this.props}>
        {this.props.options.map((option, i) => {
          return (
            <OptionStripOption
              key={`options-strip-item-` + option.value}
              selected={option.value === this.state.value}
            >
              <input
                type='radio'
                name={`options-strip-${this.props.options[0].label}`}
                onChange={this.handleChange}
                value={option.value}
                checked={
                  option.value === this.state.value ? true : false
                }
              />
              <Text size='small'>
                {option.label ? option.label : option.value}
              </Text>
            </OptionStripOption>
          )
        })}
      </OptionStripContainer>
    )
  }

  public componentDidUpdate (prevState: any) {
    if (this.props.value && prevState.value !== this.props.value) {
      this.setState({value: this.props.value})
    }
  }

  private updateState(value:string, event?:any) {
    this.setState(
      {
        value: value
      },
      () => {
        if (this.props.onChange) {
          this.props.onChange(event)
        }
      }
    )
  }

  private handleChange (event: React.ChangeEvent<HTMLInputElement>) {
    event.persist()
    const value = event.target.value
    const storedEvent = event
    this.updateState(value, storedEvent)
  }
}

export const OptionStrip = styled(OptionStripFactory)``
