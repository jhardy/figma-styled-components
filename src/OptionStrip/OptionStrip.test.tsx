import 'jest-styled-components'
import * as React from 'react'
import * as renderer from 'react-test-renderer'
import { OptionStrip } from './OptionStrip'


test('A default option strip', () => {
  const tree = renderer.create(<OptionStrip options={[{ value: '1', label: 'One' }, { value: '2', label: 'Two' } ]} />).toJSON()
  expect(tree).toMatchSnapshot()
})

test('A option strip with default value set', () => {
  const tree = renderer.create(<OptionStrip defaultSelected={1} options={[{ value: '1', label: 'One' }, { value: '2', label: 'Two' } ]} />).toJSON()
  expect(tree).toMatchSnapshot()
})

