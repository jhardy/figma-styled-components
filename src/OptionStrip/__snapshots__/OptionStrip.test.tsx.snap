// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`A default option strip 1`] = `
.c3 {
  box-sizing: border-box;
  margin: 0;
  padding: 0;
  font-family: "Inter",sans-serif;
  color: rgba(0,0,0,0.8);
  font-weight: 400;
  font-size: 11px;
  -webkit-letter-spacing: 0.005em;
  -moz-letter-spacing: 0.005em;
  -ms-letter-spacing: 0.005em;
  letter-spacing: 0.005em;
  line-height: 16px;
}

.c0 {
  box-sizing: border-box;
  border-radius: 2px;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  height: 24px;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c0:hover {
  box-shadow: inset 0 0 0 1px rgba(0,0,0,0.1);
}

.c1 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background: rgba(0,0,0,0.1);
  box-sizing: border-box;
  border-radius: 2px;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  height: 24px;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  min-width: 24px;
  padding: 5px;
  position: relative;
}

.c1 .c2 {
  line-height: 0;
}

.c1 input {
  position: absolute;
  width: 100%;
  height: 100%;
  opacity: 0;
}

.c4 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background: transparent;
  box-sizing: border-box;
  border-radius: 2px;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  height: 24px;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  min-width: 24px;
  padding: 5px;
  position: relative;
}

.c4 .c2 {
  line-height: 0;
}

.c4 input {
  position: absolute;
  width: 100%;
  height: 100%;
  opacity: 0;
}

<div
  className="c0"
>
  <div
    className="c1"
    selected={true}
  >
    <input
      checked={true}
      name="options-strip-One"
      onChange={[Function]}
      type="radio"
      value="1"
    />
    <span
      className="c2 c3"
      size="small"
    >
      One
    </span>
  </div>
  <div
    className="c4"
    selected={false}
  >
    <input
      checked={false}
      name="options-strip-One"
      onChange={[Function]}
      type="radio"
      value="2"
    />
    <span
      className="c2 c3"
      size="small"
    >
      Two
    </span>
  </div>
</div>
`;

exports[`A option strip with default value set 1`] = `
.c3 {
  box-sizing: border-box;
  margin: 0;
  padding: 0;
  font-family: "Inter",sans-serif;
  color: rgba(0,0,0,0.8);
  font-weight: 400;
  font-size: 11px;
  -webkit-letter-spacing: 0.005em;
  -moz-letter-spacing: 0.005em;
  -ms-letter-spacing: 0.005em;
  letter-spacing: 0.005em;
  line-height: 16px;
}

.c0 {
  box-sizing: border-box;
  border-radius: 2px;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  height: 24px;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c0:hover {
  box-shadow: inset 0 0 0 1px rgba(0,0,0,0.1);
}

.c1 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background: rgba(0,0,0,0.1);
  box-sizing: border-box;
  border-radius: 2px;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  height: 24px;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  min-width: 24px;
  padding: 5px;
  position: relative;
}

.c1 .c2 {
  line-height: 0;
}

.c1 input {
  position: absolute;
  width: 100%;
  height: 100%;
  opacity: 0;
}

.c4 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background: transparent;
  box-sizing: border-box;
  border-radius: 2px;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  height: 24px;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  min-width: 24px;
  padding: 5px;
  position: relative;
}

.c4 .c2 {
  line-height: 0;
}

.c4 input {
  position: absolute;
  width: 100%;
  height: 100%;
  opacity: 0;
}

<div
  className="c0"
  defaultValue="1"
>
  <div
    className="c1"
    selected={true}
  >
    <input
      checked={true}
      name="options-strip-One"
      onChange={[Function]}
      type="radio"
      value="1"
    />
    <span
      className="c2 c3"
      size="small"
    >
      One
    </span>
  </div>
  <div
    className="c4"
    selected={false}
  >
    <input
      checked={false}
      name="options-strip-One"
      onChange={[Function]}
      type="radio"
      value="2"
    />
    <span
      className="c2 c3"
      size="small"
    >
      Two
    </span>
  </div>
</div>
`;
